### ISSUE TITLE
Issue with frequent disconnection

### ISSUE THREAD

**Issue by waffleau** [kafka4beam/brod#312]
Hi,

I've recently been having issues with my Kafka connection dropping out roughly every five seconds. I've been using brod with this cluster for about six months, and these issues are fairly new, but I'm not exactly sure where the fault is.

I spent a lot of time yesterday trying to make sure the cluster itself wasn't the issue. Using`kafka-console-consumer` I can connect and receive messages without any issues, so I think something is wrong on the Elixir side, but I'm not sure where things are going wrong.

OTP version:
```
Erlang/OTP 21 [erts-10.3] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe] [dtrace]

Elixir 1.8.1 (compiled with Erlang/OTP 21)
```

I'm using brod 3.7.6 (previously 3.7.2), and I've tried several other versions with no luck. Here is a mix task which replicates the problem:

```elixir
defmodule Mix.Tasks.Kafka.Test do
  use Mix.Task

  def run(_) do
    Application.ensure_all_started(:brod)

    bootstrap = [{"my.kafka.host", 9092}]
    client = :my_kafka_client
    topic = "my-topic"

    :ok =
      :brod.start_client(
        bootstrap,
        client,
        [
          {:sasl, {:scram_sha_256, "username", "password"}},
          {:ssl, [{:cacertfile, "cert.pem"}]}
        ]
      )

    :ok = :brod.start_producer(client, topic, [])

    callback_message = fn _partition, msg, state ->
      IO.inspect(msg)

      {:ok, :ack, state}
    end

    callback_state = fn _ -> nil end

    {:ok, _pid} =
      :brod_topic_subscriber.start_link(
        client,
        topic,
        :all,
        [{:begin_offset, :latest}],
        [],
        :message,
        callback_message,
        callback_state
      )

    Process.sleep(2000)

    :ok = :brod.produce_sync(client, topic, 0, "key1", "hello")

    Process.sleep(10000)

    :ok = :brod.produce_sync(client, topic, 0, "key2", "goodbye")

    Process.sleep(10000)
  end
end
```

Running the above example, this is the output (scrubbed some personal details):
```
18:25:40.407 [info]  [supervisor: {:local, :brod_sup}, started: [pid: #PID<0.129.0>, id: :my_kafka_client, mfargs: {:brod_client, :start_link, [[{"my.kafka.host", 9092}], :my_kafka_client, [sasl: {:scram_sha_256, "username", #Function<7.4423944/0 in :brod_utils.init_sasl_opt/1>}, ssl: [cacertfile: "cert.pem"]]]}, restart_type: {:permanent, 10}, shutdown: 5000, child_type: :worker]]

18:25:40.687 [info]  [supervisor: {#PID<0.137.0>, :brod_producers_sup}, started: [pid: #PID<0.139.0>, id: "my-topic", mfargs: {:supervisor3, :start_link, [:brod_producers_sup, {:brod_producers_sup2, #PID<0.129.0>, "my-topic", []}]}, restart_type: {:permanent, 10}, shutdown: :infinity, child_type: :supervisor]]

18:25:40.690 [info]  [supervisor: {#PID<0.138.0>, :brod_consumers_sup}, started: [pid: #PID<0.141.0>, id: "my-topic", mfargs: {:supervisor3, :start_link, [:brod_consumers_sup, {:brod_consumers_sup2, #PID<0.129.0>, "my-topic", [begin_offset: :latest]}]}, restart_type: {:permanent, 10}, shutdown: :infinity, child_type: :supervisor]]

18:25:40.696 [info]  [supervisor: {#PID<0.141.0>, :brod_consumers_sup}, started: [pid: #PID<0.143.0>, id: 0, mfargs: {:brod_consumer, :start_link, [#PID<0.129.0>, "my-topic", 0, [begin_offset: :latest]]}, restart_type: {:transient, 2}, shutdown: 5000, child_type: :worker]]

18:25:40.698 [info]  [supervisor: {#PID<0.139.0>, :brod_producers_sup}, started: [pid: #PID<0.142.0>, id: 0, mfargs: {:brod_producer, :start_link, [#PID<0.129.0>, "my-topic", 0, []]}, restart_type: {:permanent, 5}, shutdown: 5000, child_type: :worker]]

18:25:40.899 [info]  client :my_kafka_client connected to 1.2.3.4:9092

{:kafka_message, 75, "key1", "hello", :create, 1552721142691, []}

18:25:46.058 [info]  client :my_kafka_client: payload connection down 1.2.3.4:9092
reason:{:send_error, :closed}

18:25:48.420 [info]  client :my_kafka_client connected to 1.2.3.4:9092

18:25:53.523 [info]  client :my_kafka_client: payload connection down 1.2.3.4:9092
reason:{:send_error, :closed}

18:25:55.869 [info]  client :my_kafka_client connected to 1.2.4.4:9092

{:kafka_message, 76, "key2", "goodbye", :create, 1552721152742, []}

18:26:00.982 [info]  client :my_kafka_client: payload connection down 1.2.3.4:9092
reason:{:send_error, :closed}
```

Any help that you can provide would be much appreciated, I've been banging my head against this issue for a couple of days.

**Comment by zmstone** [comment]
Which kafka version?
In your logs, the interval is not 5 seconds, but 6~7.
This make me wonder if the kafka config `connections.max.idle.ms` is set to 6 seconds.
default is 600 seconds (10 minutes).

**Comment by waffleau** [comment]
The Kafka version is 1.1.0. We use a managed cluster via Instaclustr, so I'm not sure if what the idle timeout is. Is this something I can query from Kafka? Otherwise I'll open a support ticket with them to find out.

**Comment by zmstone** [comment]
No, you can not query this config, it's a broker config, not a topic config.
You can test it locally to reproduce it first if you like.

**Comment by waffleau** [comment]
I'm just spun a Kafka instance via Docker and I'm not having the issue, so I guess it's something configuration based on the service provider. I'll open a ticket with them. Closing this for now, if that turns out not to be the problem I'll reopen.

Thanks for your help!

**Comment by waffleau** [comment]
@zmstone as an update to this in case anyone else opens up issues, OTP 21.3 was our problem: https://github.com/erlang/otp/releases/tag/OTP-21.3. The changes to SSL didn't play well with this library. We fixed it by rolling back to 21.2.7.
